homeassistant/components/formula_one/coordinator.py

```diff
@@ -0,0 +1,192 @@
+        self.races: list[ergast.Race] = []
+        self.races_next_year: list[ergast.Race] = []
+        self.season = -1
+        self.round = -1
+        self.entity_ids: list[str] = []
+
+    async def _async_update_data(self) -> None:
+        """Fetch data from API endpoint."""
+
+        async with asyncio.timeout(60):
+            await self.update()
+
+    async def update(self):
+        """Update status from the online service."""
+        _LOGGER.debug("Updating")
+
+        try:
+            constructor_standings, driver_standings, races = await asyncio.gather(
+                self.hass.async_add_executor_job(
+                    self.ergast.season().get_constructor_standings
+                ),
+                self.hass.async_add_executor_job(
+                    self.ergast.season().get_driver_standings
+                ),
+                self.hass.async_add_executor_job(self.ergast.season().get_races),
+            )
+
+            current_season = int(races[0].season)
+            next_season = current_season + 1
+
+            races_next_year = await self.hass.async_add_executor_job(
+                self.ergast.season(next_season).get_races
+            )
+
+        except JSONDecodeError as ex:
+            raise UpdateFailed(ex) from ex
+        except RequestsExceptions.RequestException as ex:
+            raise UpdateFailed(ex) from ex
+        except Exception as ex:
+            raise UpdateFailed(ex) from ex
+
+        if (
+            len(constructor_standings) != 1
+            or len(driver_standings) != 1
+            or len(races) < 1
+        ):
+            raise UpdateFailed(
+                "The Ergast server returned an unexpected number of results"
+            )
+
+        new_sensors_discovered = self.new_sensors_discovered(
+            constructor_standings[0].constructor_standings,
+            driver_standings[0].driver_standings,
+            races,
+        )
+
+        self.constructor_standings = constructor_standings[0].constructor_standings
+        self.driver_standings = driver_standings[0].driver_standings
+        self.races = races
+        self.races_next_year = races_next_year
+        self.season = constructor_standings[0].season
+        self.round = constructor_standings[0].round_no
+
+        if new_sensors_discovered:
+            async_dispatcher_send(self.hass, F1_DISCOVERY_NEW)
+
+    def get_driver_name(self, driver_id: str) -> str | None:
+        """Get the formatted name of the Driver."""
+        for standing in self.driver_standings:
+            if standing.driver.driver_id == driver_id:
+                return f"{standing.driver.given_name} {standing.driver.family_name}"
+
+        return None
+
+    def new_sensors_discovered(
+        self,
+        new_constructor_standings: list[ergast.ConstructorStanding],
+        new_driver_standings: list[ergast.DriverStanding],
+        new_races: list[ergast.Race],
+    ) -> bool:
```

Comment:
This function is not needed and just adds a risk of bugs.
Instead, in `sensor.async_setup_entry` call `coordinator.async_add_listener(async_setup_sensors)`